KAFKA-17584: Fix incorrect synonym handling for dynamic log configurations

cmccabe opened the pull request:
Several Kafka log configurations in have synonyms. For example, log retention can be configured either by log.retention.ms, or by log.retention.minutes, or by log.retention.hours. There is also a faculty in Kafka to dynamically change broker configurations without restarting the broker. These dynamically set configurations are stored in the metadata log and override what is in the broker properties file.

Unfortunately, these two features interacted poorly; there was a bug where the dynamic log configuration update code ignored synonyms. For example, if you set log.retention.minutes and then reconfigured something unrelated that triggered the LogConfig update path, the retention value that you had configured was overwritten.

The reason for this was incorrect handling of synonyms. The code tried to treat the Kafka broker configuration as a bag of key/value entities rather than extracting the correct retention time (or other setting with overrides) from the KafkaConfig object.

cmccabe committed:
KAFKA-17584: Fix incorrect synonym handling for dynamic log configurations

Several Kafka log configurations in have synonyms. For example, log retention can be configured
either by log.retention.ms, or by log.retention.minutes, or by log.retention.hours. There is also
a faculty in Kafka to dynamically change broker configurations without restarting the broker. These
dynamically set configurations are stored in the metadata log and override what is in the broker
properties file.

Unfortunately, these two features interacted poorly; there was a bug where the dynamic log
configuration update code ignored synonyms. For example, if you set log.retention.minutes and then
reconfigured something unrelated that triggered the LogConfig update path, the retention value that
you had configured was overwritten.

The reason for this was incorrect handling of synonyms. The code tried to treat the Kafka broker
configuration as a bag of key/value entities rather than extracting the correct retention time (or
other setting with overrides) from the KafkaConfig object.

Separately from the above bug, the code did not honor the value of dynamically configured synonyms:
setting log.retention.minutes had no effect; only log.retention.ms was honored.

showuon commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -662,12 +662,27 @@ trait BrokerReconfigurable {
+    results.asScala
+  }
+
+  val KafkaConfigToLogConfigName: Map[String, String] = {
```
nit: After this PR, this variable is only used in `DynamicBrokerReconfigurationTest`. We can move it there.

showuon reviewed (approved):
LGTM! Minor comment left. 
There's a `DynamicBrokerReconfigurationTest.testConfigDescribeUsingAdminClient` test failed. I had a look, it looks like we were validating the wrong thing before. The `log.retention.hours` should be not a read-only config. It can be dynamically changed, as well as `log.roll.hours`.

fvaleri commented:
> The log.retention.hours should be not a read-only config. It can be dynamically changed, as well as log.roll.hours.

I think we should also update the documentation, as they are both reported as read-only.

jayteej commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -662,12 +662,27 @@ trait BrokerReconfigurable {
-    ServerTopicConfigSynonyms.TOPIC_CONFIG_SYNONYMS.values.asScala.toSet - ServerLogConfigs.LOG_MESSAGE_FORMAT_VERSION_CONFIG
-  val KafkaConfigToLogConfigName: Map[String, String] =
-    ServerTopicConfigSynonyms.TOPIC_CONFIG_SYNONYMS.asScala.map { case (k, v) => (v, k) }
+  val ReconfigurableConfigs: Set[String] = {
```
nit (immutable style):
```
  val ReconfigurableConfigs: Set[String] = {
    ServerTopicConfigSynonyms.ALL_TOPIC_CONFIG_SYNONYMS
      .values()
      .asScala
      .flatMap(v => v.asScala.map(configSynonym => configSynonym.name()))
      .filterNot(_ == ServerLogConfigs.LOG_MESSAGE_FORMAT_VERSION_CONFIG)
      .toSet
  }
```

showuon commented:
> > The log.retention.hours should be not a read-only config. It can be dynamically changed, as well as log.roll.hours.
> 
> I think we should also update the documentation, as they are both reported as read-only.

Good point! But the doc is actually generated, and I just confirmed, after this PR, the doc will be correctly updated:

```
log.retention.hours

The number of hours to keep a log file before deleting it (in hours), tertiary to log.retention.ms property
Type:	int
Default:	168
Valid Values:	
Importance:	high
Update Mode:	cluster-wide
log.retention.minutes

The number of minutes to keep a log file before deleting it (in minutes), secondary to log.retention.ms property. If not set, the value in log.retention.hours is used
Type:	int
Default:	null
Valid Values:	
Importance:	high
Update Mode:	cluster-wide
log.retention.ms

The number of milliseconds to keep a log file before deleting it (in milliseconds), If not set, the value in log.retention.minutes is used. If set to -1, no time limit is applied.
Type:	long
Default:	null
Valid Values:	
Importance:	high
Update Mode:	cluster-wide
log.roll.hours

The maximum time before a new log segment is rolled out (in hours), secondary to log.roll.ms property
Type:	int
Default:	168
Valid Values:	[1,...]
Importance:	high
Update Mode:	cluster-wide
log.roll.jitter.hours

The maximum jitter to subtract from logRollTimeMillis (in hours), secondary to log.roll.jitter.ms property
Type:	int
Default:	0
Valid Values:	[0,...]
Importance:	high
Update Mode:	cluster-wide
```

jayteej commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -662,12 +662,27 @@ trait BrokerReconfigurable {
+
+  val KafkaConfigToLogConfigName: Map[String, String] = {
+    val results = new util.HashMap[String, String]
+    ServerTopicConfigSynonyms.ALL_TOPIC_CONFIG_SYNONYMS.entrySet().forEach(e => {
```
nit: As per above, use flatmap and toSet

clolov reviewed (approved):
> > > The log.retention.hours should be not a read-only config. It can be dynamically changed, as well as log.roll.hours.
> > 
> > I think we should also update the documentation, as they are both reported as read-only.
> 
> Good point! But the doc is actually generated, and I just confirmed, after this PR, the doc will be correctly updated

My only call out is that currently (in trunk) log.retention.hours/minutes are documented as read-only and behave as read-only. After this change they become cluster-wide in both documentation and behaviour. I am happy with this change as long as we publicly call it out in some form.

showuon commented:
I've sent a mail dev/user mailing list to discuss it [here](https://lists.apache.org/thread/pn52rf3hrlz2fowftpfgp27o7o34h2kp). FYI.

fvaleri reviewed (commented):
Thanks @showuon.

mumrah commented:
Looks like there is a failing test:

> DynamicBrokerReconfigurationTest > "testConfigDescribeUsingAdminClient(String)

cmccabe committed:
Go back to the old behavior where synonyms cannot be used for dynamic configs

cmccabe commented:
For now, I'm going to change it back to the current behavior where we ignore the value of dynamically configured synonyms such as `log.retention.minutes`, etc. We can make that change in a separate PR if people feel strongly about it. (The fact that it's been like this since 3.0 seems like a good reason to leave it for now....)

cmccabe committed:
remove testDynamicLogReconfigurableConfigsIncludesDeprecatedSynonyms

showuon reviewed (approved):
LGTM! It's fine we keep the old behavior since it won't block any operation.

kamalcph commented:
This issue was first reported in KAFKA-15266 and a PR #14119 is opened for it.

amangandhi94 commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -732,17 +732,7 @@ class DynamicLogConfig(logManager: LogManager, server: KafkaBroker) extends Brok
   override def reconfigure(oldConfig: KafkaConfig, newConfig: KafkaConfig): Unit = {
     val originalLogConfig = logManager.currentDefaultConfig
     val originalUncleanLeaderElectionEnable = originalLogConfig.uncleanLeaderElectionEnable
-    val newBrokerDefaults = new util.HashMap[String, Object](originalLogConfig.originals)
```
Just curious - Do we know why this logic was even added? I cant seem to understand if there was any benefit to doing this.

cmccabe replied:
See discussion here https://github.com/apache/kafka/pull/17258#discussion_r1775665743

amangandhi94 reviewed (approved):
This looks good to me and addresses the bug I had raised - https://issues.apache.org/jira/browse/KAFKA-15266

Just wondering how can I get more attention to bugs or PR i have raised next time onwards?

fvaleri commented:
> Just wondering how can I get more attention to bugs or PR i have raised next time onwards?

In this case raising the bug as a blocker, due to the potential data loss. In general, sending an email to the dev mailing list would be enough to catch some more attention in my experience. Hope it helps.

fvaleri reviewed (approved):
LGTM. Thanks.

kamalcph reviewed (approved):
LGTM

rajinisivaram commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -732,17 +732,7 @@ class DynamicLogConfig(logManager: LogManager, server: KafkaBroker) extends Brok
     val originalUncleanLeaderElectionEnable = originalLogConfig.uncleanLeaderElectionEnable
-    val newBrokerDefaults = new util.HashMap[String, Object](originalLogConfig.originals)
-    newConfig.valuesFromThisConfig.forEach { (k, v) =>
-      if (DynamicLogConfig.ReconfigurableConfigs.contains(k)) {
```
It has been a long time, so I may not remember correctly. But I think we used to only update configs which are `ReconfigurableConfigs` even if you updated ZooKeeper directly with new configs that included others. Non-reconfigurable configs would be picked up only on the next broker restart. Do we think it is safe to include all configs now either because we don't allow other configs to be updated in KRaft or because all log configs are reconfigurable now?

cmccabe replied:
Thanks for this comment, @rajinisivaram . I think you're right that we should be excluding changes to configurations that don't appear in `DynamicLogConfig.ReconfigurableConfigs`. I have added some code to do this now. I think from a practical point of view, this only affects `message.format.version`, which we explicitly excluded from being dynamically reconfigurable.

> Do we think it is safe to include all configs now either because we don't allow other configs to be updated in KRaft or because all log configs are reconfigurable now?

KRaft has the same behavior as ZK, in that it lets you set any broker configuration you want, whether it's valid or not. There were too many people depending on this for us to change it in 3.x. Maybe in the future if someone creates a KIP...

Of course the command-line tool does its own validation.

showuon replied:
Good call @rajinisivaram !

cmccabe committed:
Do not allow log configurations to be changed if they're not reconfigurable

- Add testLogRetentionTimeMinutesIsNotDynamicallyReconfigurable

- clean up some cases where we were using zkconnect but did not need to

rajinisivaram reviewed (approved):
@cmccabe Thanks for the PR, LGTM

junrao commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -732,17 +732,12 @@ class DynamicLogConfig(logManager: LogManager, server: KafkaBroker) extends Brok
+    val newBrokerDefaults = new util.HashMap[String, Object](newConfig.extractLogConfigMap)
+    originalLogConfig.originals().forEach((k, v) => {
+      if (!DynamicLogConfig.ReconfigurableConfigs.contains(k)) {
+        newBrokerDefaults.put(k, v)
```
1. Do we need to translate the name from KafkaConfig to LogConfig as the original code does?
2. We started `newBrokerDefaults` with `newConfig.extractLogConfigMap` that includes non-reconfigurable configs. Where is the logic to remove them?

cmccabe replied:
Let me fix this a bit. It should be sufficient to just remove the non-reconfigurable configs (really, singular config)

junrao reviewed (commented):
@cmccabe : Thanks for the PR. Added a couple of comments.

cmccabe committed:
rework the reconfiguration logic a bit to fix the non-reconfigurable log config

junrao commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -732,17 +744,14 @@ class DynamicLogConfig(logManager: LogManager, server: KafkaBroker) extends Brok
+    val newBrokerDefaults = new util.HashMap[String, Object](newConfig.extractLogConfigMap)
+    val originalLogConfigMap = originalLogConfig.originals()
+    DynamicLogConfig.NonReconfigrableLogConfigs.foreach(k => {
+      Option(originalLogConfigMap.get(k)) match {
```
Hmm, `originalLogConfigMap` is the current config. We should check from `newConfig`, right?

cmccabe replied:
Non-reconfigurable configs are copied over from the *current* (not new) configuration.

junrao commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -732,17 +744,14 @@ class DynamicLogConfig(logManager: LogManager, server: KafkaBroker) extends Brok
+    DynamicLogConfig.NonReconfigrableLogConfigs.foreach(k => {
+      Option(originalLogConfigMap.get(k)) match {
+        case None => newBrokerDefaults.remove(k)
+        case Some(v) => newBrokerDefaults.put(k, v)
```
Not sure that I follow here. Why are we putting a non-reconfigurable config to newBrokerDefaults?

cmccabe replied:
Because we want the non-reconfigurable configuration to have the same value (or lack of value) that it had previously. It should not change.

junrao commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -732,17 +744,14 @@ class DynamicLogConfig(logManager: LogManager, server: KafkaBroker) extends Brok
+        case Some(v) => newBrokerDefaults.put(k, v)
       }
-    }
+    })
```
`newBrokerDefaults` has the config name with log prefix, right? Should we translate the name from `KafkaConfig` to `LogConfig` as the original code does?

cmccabe replied:
> newBrokerDefaults has the config name with log prefix, right

no. it comes from `KafkaConfig.extractLogConfigMap`, which creates a map of log (aka topic) configurations, not broker configurations.

junrao reviewed (commented):
@cmccabe : Thanks for the updated PR. A few more comments.

junrao reviewed (approved):
@cmccabe : Thanks for the explanation. LGTM

showuon commented:
All tests passed.
